### ISSUE TITLE
Have you tested on any dataset?

### ISSUE THREAD

**Issue by XiaoyanLi1** [sciencefans/Focal-Loss#3]
Have you tested the focal loss on any dataset? Are there improvements with focal loss?

**Comment by liuyuisanai** [comment]
I'm now testing it on face detection task with our [RSA](https://arxiv.org/abs/1707.09531) pipeline. Since it has been claimed by the author that the result is very sensitive to the two hyper-parameters `alpha` and `gamma` , I'm still grid searching the best combination. Please stay tuned and I will report some results few days later :)

**Comment by daikankan** [comment]
for (int i = 0; i < count; ++i) {
	  if ( target[ i ] == ignore_label_ )
		  loss = 0;
	  else{
		  valid_num_+=1;
		  loss -= input_data[ i ] * ( target[ i ] - ( input_data[ i ] >= 0 ) ) -
			  log(1 + exp(input_data[ i ] - 2 * input_data[ i ] * ( input_data[ i ] >= 0 )));
	  }
}

I'm confused about if it is a bug, should it be 'loss -= 0' ？：

  for (int i = 0; i < count; ++i) {
	  if ( target[ i ] == ignore_label_ )
		  loss -= 0;
	  else{
		  valid_num_+=1;
		  loss -= input_data[ i ] * ( target[ i ] - ( input_data[ i ] >= 0 ) ) -
			  log(1 + exp(input_data[ i ] - 2 * input_data[ i ] * ( input_data[ i ] >= 0 )));
	  }
}

**Comment by XiaoyanLi1** [comment]
The CPU version is not finished. I believe the author has mentioned in README.

**Comment by XiaoyanLi1** [comment]
But I'm also confused by the formutation in .cu. It seems to be different with the paper.
1. alpha should be different for pos and neg examples
2. Is pt computed as sigmoid(input_data)? Why?

**Comment by liuyuisanai** [comment]
Hi @XiaoyanLi1 ,
For Q1: Apologize for this mistake. I've fixed this issue, thanks for your feed back!
For Q2: `pt` can be 1) sigmoid(x) or 2) softmax(x) in regard to 1) `sigmoid_cross_entropy_loss` for binary classification or 2) `softmax_loss` for multiple class (c>=3) classification. Here I implement it with the former form for fg/bg training. I may implement the latter one few days later.

**Comment by XiaoyanLi1** [comment]
@sciencefans Thanks!

**Comment by yjump** [comment]
Hi @sciencefans ,
I am working on implementing softmax focal loss by learning your excellent work these days, while I has never implemented total caffeMex as you.  I am now confused by how to implement backward_gpu. I learn that the derivation of softmax CE: when i=t, d CE/  pi =  pt -1, and i!=t d CE/d pi = pi (p for softmax(x)). but why bottom_diff[n * dim + label_value * spatial_dim + s] -=1, (n * dim + label_value * spatial_dim + s) stands for i=t? As for this sigmoid focal loss,  you implement d(oriloss)*scale in the one formultation whenever i=t or i!=t, the derivation of sigmoid focal loss is not the same i doubt

**Comment by yjump** [comment]
Hi @sciencefans , I am also confused by your implement of d(oriloss) * scaler in ,cu file line 109 to line 111. Can you kindly offer a brief deduction of derivative for sigmoid focal loss if possible? I doubt why d scaler is changed by target_value, while d(oriloss) is the same.

**Comment by liuyuisanai** [comment]
Hi @yjump , both `d(oriloss)` and `d(scaler)` are relevant with `target_value`, aka label. Here I use x for input and z for target:
`oriloss`: = z * -log(sigmoid(x)) + (1 - z) * -log(1 - sigmoid(x)) = max(x, 0) - x * z + log(1 + exp(-abs(x)))
`scaler`:= z * alpha * (1-sigmoid(x))^gamma + (1 - z) * (1 - alpha) * (sigmoid(x))^gamma
`d(oriloss)`:  = y - z
`d(scaler)`:  = -alpha * gamma * (1-(z==1?sigmoid(x):1-sigmoid(x))^(gamma-1) * exp(-abs(x))/(exp(-abs(x))+1)^2

**Comment by yjump** [comment]
@sciencefans Thx!

**Comment by zimenglan-sysu-512** [comment]
hi @sciencefans,

the `oriloss` is the normal cross-entropy loss, so i guess your `d(oriloss)` is wrong.

**Comment by liuyuisanai** [comment]
hi @zimenglan-sysu-512 ,
Sure I implement Focal-Loss in sigmoid-CE form. So d(oriloss) is just like the `bp` process of [sigmoid-CE-loss](https://github.com/sciencefans/CaffeMex_v2/blob/master/src/caffe/layers/sigmoid_cross_entropy_loss_layer.cu). Is there anything wrong?

> pt can be 1) sigmoid(x) or 2) softmax(x) in regard to 1) sigmoid_cross_entropy_loss for binary classification or 2) softmax_loss for multiple class (c>=3) classification. Here I implement it with the former form for fg/bg training. I may implement the latter one few days later.

**Comment by zimenglan-sysu-512** [comment]
hi @sciencefans,

 can you show how to get your `d(oriloss)` in details?

**Comment by liuyuisanai** [comment]
@zimenglan-sysu-512 @yjump   
Oops, sorry guys there is a typo above. I wrote down the forward inference to bp formulation by mistake.
d(oriloss) should be y - z.
Here we get input `x`, label `z` and let `y` = sigmoid(`x`),
`oriloss` = z * -log(sigmoid(x)) + (1 - z) * -log(1 - sigmoid(x)) = -xz + log(1+exp(x))
d(oriloss)/d(x) = -z + exp(x)/(1+exp(x)) = -z + y = y - z
This is implemented in Line 109 to 110 of [.cu](https://github.com/sciencefans/Focal-Loss/blob/master/focal_loss_layer.cu)

**Comment by zimenglan-sysu-512** [comment]
hi @sciencefans,

thanks for update. btw, `z` is either 1 or 0?

**Comment by liuyuisanai** [comment]
@zimenglan-sysu-512 
Yes. In original CE loss, z can be a probability between [0, 1]. However in Focal-Loss, I believe it should only be either 1 or 0 since p_t and alpha_t are determined by whether z == 1.

**Comment by zimenglan-sysu-512** [comment]
hi @sciencefans,

i am confused by `d(scaler)`.
so can u show it more simply in details?

thanks.

**Comment by liuyuisanai** [comment]
@zimenglan-sysu-512 
d(scaler): 
= alpha_t * gamma * (1-p)^(gamma-1)* -1 * d(i)/d(x)
= ...
= -alpha_t * gamma * (1-p_t^(gamma-1) * exp(-abs(x))/(exp(-abs(x))+1)^2
